agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "php_agent.h"
+#include "php_user_instrument_hashmap.h"
+
+#if ZEND_MODULE_API_NO >= ZEND_7_4_X_API_NO
+
+typedef struct _nr_wraprecs_bucket {
+  struct _nr_wraprecs_bucket* prev;
+  struct _nr_wraprecs_bucket* next;
+  /* for efficiency wraprec is used as both: key and a value */
+  nruserfn_t* wraprec;
+} nr_wraprecs_bucket_t;
+
+typedef struct _nr_php_wraprec_hashmap {
+  nr_php_wraprec_hashmap_dtor_fn_t dtor_func;
+  size_t log2_num_buckets;
+  nr_wraprecs_bucket_t** buckets;
+  size_t elements;
+} nr_php_wraprec_hashmap_t;
+
+static inline size_t nr_count_buckets(const nr_php_wraprec_hashmap_t* hashmap) {
+  return (size_t)(1 << hashmap->log2_num_buckets);
+}
```

Comment:
This is purely a style nit, but is there a reason to have these defined here as opposed to the .h file?  I think it would be slightly cleaner if all the necessary struct definitions were extracted out to the header.

Reply:
This is intentionally not exposed in .h file. This type of code structure is called [opaque data type](https://en.wikipedia.org/wiki/Opaque_data_type). The internals of hashmap implementation should stay hidden away from the interface in case they need change.